### ISSUE TITLE
iquest fails when select is used in argument string

### ISSUE THREAD

**Issue by AJdeMooij** [irods/irods_client_icommands#4]
If the character sequence "select" (in lower case) is used in an iquest command, iquest fails with status -110700 (no column names found).

```bash
$ iquest "SELECT COLL_NAME WHERE COLL_NAME = 'safe_word'"
CAT_NO_ROWS_FOUND: Nothing was found matching your query

$ iquest "SELECT COLL_NAME WHERE COLL_NAME = 'SELECT'"
CAT_NO_ROWS_FOUND: Nothing was found matching your query

$ iquest "SELECT COLL_NAME WHERE COLL_NAME = 'select'"
ERROR: iquest Error: queryAndShowStrCond failed status = -1107000 NO_COLUMN_NAME_FOUND
```

Further testing shows that escaping any character in the word 'select', except for the first, can be used as a workaround:

```bash
$ iquest "SELECT COLL_NAME WHERE COLL_NAME = '\select'"
ERROR: iquest Error: queryAndShowStrCond failed status = -1107000 NO_COLUMN_NAME_FOUND

$ iquest "SELECT COLL_NAME WHERE COLL_NAME = 's\elect'"
CAT_NO_ROWS_FOUND: Nothing was found matching your query

$ iquest "SELECT COLL_NAME WHERE COLL_NAME = 'selec\t'"
CAT_NO_ROWS_FOUND: Nothing was found matching your query
```

The problem does not occur when using general queries in irules.

**Comment by kellerb** [comment]
The [query string is parsed](https://github.com/irods/irods/blob/77dfa1120358d233451cb00dcdb2255e4ef37ebc/lib/core/src/rcMisc.cpp#L3036-L3037) by first searching (via raw `strstr()`) for `select`, and, if not found, for `SELECT`. This means that, since your query string uses the uppercase version of the keyword, the parsing code ends up finding and anchoring the rest of the parsing around the lowercase `select` substring at the end of the query string.

The workaround is to always use (when available) the lowercase versions of the genquery keywords.

Adding backslashes (except at the beginning of `select`) changes the perceived behavior of `iquest`, because you are inserting a `\` character (value `0x09`) into the `select` substring, which prevents `strstr()` from matching against it.

This issue is related in principle to https://github.com/irods/irods/issues/3064, in which the same problem was happening except with the {`in`, `IN`} keywords.

We intend on fixing this class of issues by replacing the current ad-hoc genquery parsing code with a more general parsing implementation (e.g. a flex/bison implementation, an example of which can be seen [here](https://github.com/kellerb/genquery-ast-example)).

**Comment by othmar-weber** [comment]
@kellerb 
The issue https://github.com/irods/irods/issues/3613 has been flagged for the iRODS 4.1.11 milestone, but has been closed as duplicate of this one.
Now iRODS 4.1.11 is out and the duplicate issue is still open.
On our system which is at iRODS 4.1.11 we also still see the error.

Can you please check if it should be solved in 4.1.11? 
If it is not included: Could you please flag it for the next realeases?

**Comment by trel** [comment]
I don't believe this will be a part of 4.1.x, or even 4.2.x.   Replacing genQuery's parsing code will be a pretty major change and not part of a patch release.

**Comment by othmar-weber** [comment]
@trel 
could you please put it on the list for 4.3?

**Comment by trel** [comment]
Yes, targeting 4.3.0.
